packages/trace-viewer/src/sw/snapshotRenderer.ts

```diff
@@ -236,10 +236,33 @@ function snapshotNodes(snapshot: FrameSnapshot): NodeSnapshot[] {
   return (snapshot as any)._nodes;
 }
 
-function snapshotScript(...targetIds: (string | undefined)[]) {
-  function applyPlaywrightAttributes(unwrapPopoutUrl: (url: string) => string, ...targetIds: (string | undefined)[]) {
+type ViewportSize = { width: number, height: number };
+type BoundingRect = { left: number, top: number, right: number, bottom: number };
+type CanvasRenderInfo = {
+  viewport: ViewportSize;
+  frames: WeakMap<Element, {
+    boundingRect?: BoundingRect;
+    scrollLeft: number;
+    scrollTop: number;
+  }>;
+};
+
+declare global {
+  interface Window {
+    __playwright_canvas_render_info__: CanvasRenderInfo;
+  }
+}
+
+function snapshotScript(viewport: ViewportSize, ...targetIds: (string | undefined)[]) {
+  function applyPlaywrightAttributes(unwrapPopoutUrl: (url: string) => string, viewport: ViewportSize, ...targetIds: (string | undefined)[]) {
     const isUnderTest = new URLSearchParams(location.search).has('isUnderTest');
 
+    const canvasRenderInfo = {
+      viewport,
+      frames: new WeakMap(),
+    };
+    window['__playwright_canvas_render_info__'] = canvasRenderInfo;
```

Comment:
after reading through this, my impression is that there's a non-trivial order of execution here, where `snapshotScript` is being executed for all frames in the frame, and they all read each other's `window.__playwright_canvas_render_info__` states. Is this understanding right?

Reply:
Yes, you're right. Each parent windows, before rendering each iframe, processes it and extracts the `__playwright_canvas_render_info__ ` and keeps it that variable. It then removes the attribute and renders the element normally, which will eventually trigger recursively the same process inside the iframe.

Reply:
Nice, that makes sense. Could you add a comment about that somewhere, for future readers?

Reply:
Also, would `__playwright_frame_bounding_rects__` be a more fitting name?

Reply:
What about this suggestion I gave on https://github.com/microsoft/playwright/pull/33809#discussion_r1868250259 ?

> Maybe topFrameWindow should be called topSnapshotWindow instead.

Reply:
Sounds good as well!